website/static/js/share/utils.js

```diff
@@ -190,30 +228,269 @@ utils.removeFilter = function(vm, filter){
+    ret.terms.size = 0;
+    ret.terms.exclude = 'of|and|or';
+    ret.terms.min_doc_count = min_doc_count;
+    return ret;
+};
+
+/* Creates a match query */
+utils.matchQuery = function (field, value) {
+    var ret = {'match': {}};
+    ret.match[field] = value;
+    return ret;
+};
+
+/* Creates a range filter */
+utils.rangeFilter = function (field_name, gte, lte) {
+    lte = lte || new Date().getTime();
+    gte = gte || 0;
+    var ret = {'range': {}};
+    ret.range[field_name] = {'gte': gte, 'lte': lte};
+    return ret;
+};
+
+/* Creates a bool query */
+utils.boolQuery = function (must, must_not, should, minimum) {
+    var ret = {
+        'bool': {
+            'must': (must || []),
+            'must_not': (must_not || []),
+            'should': (should || [])
+        }
+    };
+    if (minimum) {
+        ret.bool.minimum_should_match = minimum;
+    }
+
+    return ret;
+};
+
+/* Creates a date histogram filter */
+utils.dateHistogramFilter = function (field, gte, lte, interval) {
+    //gte and lte in ms since epoch
+    lte = lte || new Date().getTime();
+    gte = gte || 0;
+
+    interval = interval || 'week';
+    return {
+        'date_histogram': {
+            'field': field,
+            'interval': interval,
+            'min_doc_count': 0,
+            'extended_bounds': {
+                'min': gte,
+                'max': lte
             }
-            vm.graphs[type].load(vm.statsData.charts[type]);
-        }, utils.errorState.bind(this, vm));
-        vm.statsLoaded(true);
-    }).then(m.redraw);
+        }
+    };
+};
+
+/* Creates a common query */
+utils.commonQuery = function (query_string, field) {
+    field = field || '_all';
+    var ret = {'common': {}};
+    ret.common[field] = {
+        query: query_string
+    };
+    return ret;
+};
 
+/* Creates a match_all query */
+utils.matchAllQuery = function () {
+    return {
+        match_all: {}
+    };
 };
 
+/* Creates an and filter */
+utils.andFilter = function (filters) {
+    return {
+        'and': filters
+    };
+};
+
+/* Creates a query filter */
+utils.queryFilter = function (query) {
+    return {
+        query: query
+    };
+};
+
+/**
+ * Parses a filter string into one of the above filters
+ *
+ * @param {String} filterString A string representation of a filter dictionary
+ */
+utils.parseFilter = function (filterString) {
+    // parses a filter of the form
+    // filterName:fieldName:param1:param2...
+    // range:providerUpdatedDateTime:2015-06-05:2015-06-16
+    var parts = filterString.split(':');
+    var type = parts[0];
+    var field = parts[1];
+    if (type === 'range') {
+        return utils.rangeFilter(field, parts[2], parts[3]);
+    } else if (type === 'match') {
+        return utils.queryFilter(
+            utils.matchQuery(field, parts[2])
+        );
+    }
+    // Any time you add a filter, put it here
+    // TODO: Maybe this would be better as a map?
+};
+
+utils.processStats = function (vm, data) {
```

Comment:
This function mixes concerns:

1. Processing data (updating `vm` based on `data`; no browser/UI needed)
2. Displaying an error in the UI (growl)

There are a few ways to decouple. Here are some sci-fi examples:

- Exception handling

```js
// processStats no longer calls $osf.growl itself
try {
    utils.processStats(...)
} catch (err) {
    $osf.growl(...)
}
```

- Promise

```
// Perhaps we even refactor so that we don't need to modify vm inplace?
utils.processStats(...).then(
    function(updatedData) { // do stuff with data },
    function(err) { $osf.growl(...); }
)
```

- Ensure that `processStats` is always passed valid data

```javascript
if (data.aggregations) { 
    utils.processStats(vm, data);
} else {
    $osf.growl(...)
}
```